lib/internal/modules/esm/importmap.js

```diff
@@ -0,0 +1,135 @@
+'use strict';
+const path = require('path');
+const { readFileSync } = require('fs');
+const { pathToFileURL, isURL } = require('internal/url');
+const { setOwnProperty } = require('internal/util');
+const { JSONParse, ObjectEntries, SafeMap } = primordials;
+// @TODO add error codes
+// const { } = require('internal/errors').codes;
+
+class ImportMap {
+  #baseURL;
+  imports = new SafeMap();
+  scopes = new SafeMap();
+
+  constructor(raw, baseURL) {
+    this.#baseURL = baseURL;
+    processImportMap(this, baseURL, raw);
+  }
+
+  get baseURL() {
+    return this.#baseURL;
+  }
+
+  static load(filepath) {
+    let str;
+    try {
+      str = readFileSync(filepath);
+    } catch (err) {
+      if (err.code !== 'ENOENT') {
+        setOwnProperty(err, 'importMapPath', `${filepath}`);
+        throw err;
+      }
+      return;
+    }
+
+    let json;
+    try {
+      json = JSONParse(str);
+    } catch (err) {
+      setOwnProperty(err, 'importMap', `${str}`);
+      throw err;
+    }
+
+    return new ImportMap(json, pathToFileURL(path.dirname(filepath)));
+  }
+
+  resolve(specifier, parentURL) {
+    const scope = parentURL || this.baseURL;
+
+    // Process scopes
+    // @TODO this almost certainly needs to be a radix tree
+    for (const [prefix, mapping] of this.scopes) {
```

Comment:
Array destructuring uses an iterator, which is vulnerable to prototype pollution, so we gotta use indexical object destructuring instead:

```suggestion
    for (const { 0: prefix, 1: mapping } of this.scopes) {
```

Bonus: it's more efficient too.